[Diagnostics] Refactor diagnoseAmbiguityWithFixes.

hborla opened the pull request:
This change refactors `diagnoseAmbiguityWithFixes` to have a generalized strategy based on the differences between solutions in order to find the ambiguous overload, or other source of ambiguity. The new strategy will:
* Aggregate fixes from all of the solutions, grouping "common" ones based on the `FixKind` and locator.
* Identify the ambiguous overload based on the callee anchor of the overload choice and the anchor of the aggregated fixes.
* If there is no ambiguous overload, it will diagnose common fixes that appear in all of the solutions.

This change also introduces a new method on `CSFix` called `diagnoseForAmbiguity`. This is called when there is no ambiguous overload, but the fix appears in each of the solutions.

Resolves: rdar://problem/55605863

hborla committed:
[Diagnostics] Refactor diagnoseAmbigutiyWithFixes to use the differences between
solutions in order to figure out the source of ambiguity.

hborla committed:
[Diagnostics] Simplify diagnoseAmbiguityWithFixes by removing AmbiguityKind.

hborla committed:
[Diagnostics] Add CSFix::diagnoseForAmbiguity for diagnosing common
fixes that appear in all solutions.

hborla committed:
[Diagnostics] Simplify logic in diagnoseAmbiguityWithFixes for emitting
the ambiguity diagnostic by combining a few special errors into one.

hborla committed:
[Diagnostics] When diagnosing an ambiguous overload, don't mention "call"
if the expression is not a function application.

hborla committed:
[CSDiag] Remove obsolete ambiguity code.

hborla commented:
This will conflict with #29734 once merged

xedin commented on lib/Sema/CSFix.cpp:
```diff
@@ -475,6 +475,24 @@ bool DefineMemberBasedOnUse::diagnose(bool asNote) const {
+bool
+DefineMemberBasedOnUse::diagnoseForAmbiguity(ArrayRef<Solution> solutions) const {
+  Type concreteBaseType;
+  for (const auto &solution: solutions) {
```
Nit: I think this could be refactored into `llvm::all_of` call.

xedin commented on lib/Sema/ConstraintSystem.h:
```diff
@@ -837,9 +837,21 @@ class Solution {
+  using AggregatedFixesMap = llvm::SmallMapVector<ConstraintLocator *,
+      llvm::SmallMapVector<FixKind, ConstraintFixVector, 4>, 4>;
+
+  AggregatedFixesMap getAggregatedFixes() const {
```
This temporary aggregation doesn't seem essential, `diagnoseAmbiguityWithFixes` should be able to do this in one go...

xedin reviewed (commented):
Looks good so far!

slavapestov commented on test/type/protocol_composition.swift:
```diff
@@ -173,9 +173,7 @@ takesP1AndP2([Swift.AnyObject & P1 & P2]())
-// TODO(diagnostics): The problem here is that `&` is interpreted as a binary operator, we need to re-think
-// how "missing member" fix is implemented because currently it finds N solutions with multiple fixes.
-takesP1AndP2([Swift.DoesNotExist & P1 & P2]()) // expected-error {{cannot invoke '' with no arguments}}
+takesP1AndP2([Swift.DoesNotExist & P1 & P2]()) // expected-error {{module 'Swift' has no member named 'DoesNotExist'}}
```
Nice!

slavapestov commented on test/Constraints/closures.swift:
```diff
@@ -186,7 +186,7 @@ func testMap() {
 
 // <rdar://problem/22414757> "UnresolvedDot" "in wrong phase" assertion from verifier
-[].reduce { $0 + $1 }  // expected-error {{cannot invoke 'reduce' with an argument list of type '(_)'}}
+[].reduce { $0 + $1 }  // expected-error {{missing argument for parameter #1 in call}}
```
An argument list of type... what?!

Nice improvement!

xedin committed:
[Diagnostics] Diagnose ambiguity with conflicting arguments to generic parameters

It's done by first retrieving all generic parameters from each solution,
filtering boundings into distrinct set and diagnosing any differences.

For example:

```swift
func foo<T>(_: T, _: T) {}
func bar(x: Int, y: Float) {
  foo(x, y)
}
```

xedin committed:
[Diagnostics] Prioritize contextual type mismatch other ambiguities

xedin committed:
[ConstraintSystem] NFC: Remove obsolete code related to fixes from Solution

xedin commented:
@hborla I have left all of your commits intact (expect for changes related to contextual result vs. operator ref), added `diagnoseConflictingGenericArguments` which handles situations like `min(Int(0), Float(1))`. Please take a look.

xedin committed:
[Diagnostics] NFC: Rename `diagnoseConflictingArguments` into `diagnoseConflictingGenericArguments`

xedin commented:
@swift-ci please smoke test

LucianoPAlmeida commented on test/Generics/deduction.swift:
```diff
@@ -67,8 +66,9 @@ func takeTuples<T, U>(_: (T, U), _: (U, T)) {
-  takeTuples((x, y), (x, y)) // expected-error{{cannot convert value of type '(Int, Float)' to expected argument type '(Float, Int)'}}
-
+  takeTuples((x, y), (x, y))
+  // expected-error@-1 {{conflicting arguments to generic parameter 'U' ('Float' vs. 'Int')}}
```
Nice!

hborla commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -2729,6 +2729,154 @@ static void diagnoseOperatorAmbiguity(ConstraintSystem &cs,
+/// }
+///
+/// It's done by first retrieving all generic parameters from each solution,
+/// filtering boundings into distrinct set and diagnosing any differences.
```
Typos: boundings->bindings and distrinct->distinct?

xedin replied:
Fixed, thank you!

hborla commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -2729,6 +2729,154 @@ static void diagnoseOperatorAmbiguity(ConstraintSystem &cs,
+          if (auto *opaque =
+                  binding->second->getAs<OpaqueTypeArchetypeType>()) {
+            auto *decl = opaque->getDecl();
+            arguments.remove_if([&](Type argType) -> bool {
```
Could you add a short comment for why this is necessary? Took me a second to figure it out!

xedin replied:
Done!

hborla reviewed (commented):
@xedin I left 2 small comments, but otherwise your changes look great - especially the new "conflicting generic args" error message. Thank you!

LucianoPAlmeida commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -2749,239 +2897,130 @@ bool ConstraintSystem::diagnoseAmbiguityWithFixes(
+                                commonCalleeLocator);
+      return true;
+    } else {
+      bool isApplication = llvm::find_if(ArgumentInfos, [&](const auto argInfo) {
```
More of a question, since the value of `llvm::find_if` seems not to be used, just transformed into a bool with `!= ArgumentInfos.end()`, would be an option to use something `llvm::any_of` which will already return a bool?

hborla replied:
You're right, `llvm::any_of` would be better here. Thank you!

@xedin do you want to make this change or should I?

xedin replied:
I will fix it together with all my typos a bit later today :)

xedin replied:
Fixed!

LucianoPAlmeida commented on test/type/opaque.swift:
```diff
@@ -270,10 +270,15 @@ func associatedTypeIdentity() {
   sameType(gary(candace()).r_out(), gary(candace()).r_out())
   sameType(gary(doug()).r_out(), gary(doug()).r_out())
-  sameType(gary(doug()).r_out(), gary(candace()).r_out()) // expected-error {{cannot convert value of type 'some R' (result of 'candace()') to expected argument type 'some R' (result of 'doug()')}}
+  // TODO(diagnostics): This is not great but we the problem comes from the way solver discovers and attempts bindings, if we could detect that
```
Typo on `we the problem`? :)

xedin replied:
Fixed!

xedin committed:
[ConstraintSystem] NFC: Fix a couple of typos in comments

xedin committed:
[Diagnostics] Replace `llvm::find_if` with `llvm::any_of` in `diagnoseAmbiguityWithFixes`

xedin committed:
[Diagnostics] NFC: Add a comment about special handling of opaque result types

xedin commented:
@swift-ci please smoke test
